// x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/WaterfallItem.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const eui_1 = require("@elastic/eui");
const variables_1 = require("../../../../../../style/variables");
const ItemBar = styled_components_1.default('div') `
  box-sizing: border-box;
  position: relative;
  height: ${variables_1.px(variables_1.unit)};
  min-width: 2px;
  background-color: ${props => props.color};
`;
// Note: "direction: rtl;" is here to prevent text from running off of
// the right edge and instead pushing it to the left. For an example of
// how this works, see here: https://codepen.io/sqren/pen/JrXNjY
const SpanLabel = styled_components_1.default.div `
  white-space: nowrap;
  position: relative;
  direction: rtl;
  text-align: left;
  margin: ${variables_1.px(variables_1.units.quarter)} 0 0;
  font-family: ${variables_1.fontFamilyCode};
  font-size: ${variables_1.fontSizes.small};
`;
const TransactionLabel = styled_components_1.default(SpanLabel) `
  font-weight: 600;
  font-family: ${variables_1.fontFamily};
  font-size: ${variables_1.fontSize};
`;
const Container = styled_components_1.default('div') `
  position: relative;
  display: block;
  user-select: none;
  padding: ${variables_1.px(variables_1.units.half)} ${props => variables_1.px(props.timelineMargins.right)}
    ${props => variables_1.px(props.item.docType === 'span' ? variables_1.units.half : variables_1.units.quarter)}
    ${props => variables_1.px(props.timelineMargins.left)};
  border-top: 1px solid ${variables_1.colors.gray4};
  background-color: ${props => (props.isSelected ? variables_1.colors.gray5 : 'initial')};
  cursor: pointer;
  &:hover {
    background-color: ${variables_1.colors.gray5};
  }
`;
function Prefix({ item }) {
    if (item.docType !== 'transaction') {
        return null;
    }
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(eui_1.EuiIcon, { type: "merge" }),
        ' '));
}
function WaterfallItem({ timelineMargins, totalDuration, item, color, isSelected, onClick }) {
    if (!totalDuration) {
        return null;
    }
    const width = (item.duration / totalDuration) * 100;
    const left = ((item.offset + item.skew) / totalDuration) * 100;
    const Label = item.docType === 'transaction' ? TransactionLabel : SpanLabel;
    // Note: the <Prefix> appears *after* the item name in the DOM order
    // because this label is styled with "direction: rtl;" so that the name
    // itself doesn't flow outside the box to the right.
    return (react_1.default.createElement(Container, { item: item, timelineMargins: timelineMargins, isSelected: isSelected, onClick: onClick },
        react_1.default.createElement(ItemBar
        // using inline styles instead of props to avoid generating a css class for each item
        , { 
            // using inline styles instead of props to avoid generating a css class for each item
            style: { left: `${left}%`, width: `${width}%` }, color: color, type: item.docType }),
        react_1.default.createElement(Label
        // using inline styles instead of props to avoid generating a css class for each item
        , { 
            // using inline styles instead of props to avoid generating a css class for each item
            style: { left: `${left}%`, width: `${Math.max(100 - left, 0)}%` } },
            item.name,
            " ",
            react_1.default.createElement(Prefix, { item: item }))));
}
exports.WaterfallItem = WaterfallItem;
